tests/conftest.py

```diff
@@ -0,0 +1,139 @@
+    uses the `sqlalchemy_utils` library to create the database and takes a DB URL
+    as the input
+
+    Args:
+        db_url (str): Database URL
+    """
+    try:
+        logger.debug(f"Checking if database exists: {db_url.database}")
+        if not database_exists(db_url):
+            logger.info(f"Creating test database: {db_url.database}")
+            create_database(db_url)
+            logger.info(f"Test database created successfully: {db_url.database}")
+        else:
+            logger.info(f"Database already exists: {db_url.database}")
+    except Exception as e:
+        logger.error(f"Error creating database: {e}")
+        raise
+
+
+async def setup_test_database(db_url):
+    """Helper function to setup the test database
+    takes a DB URL as input and returns a SQLAlchemy engine
+
+    Args:
+        db_url (str): Database URL
+
+    Returns:
+        engine: SQLAlchemy engine
+    """
+    engine = create_async_engine(str(db_url))
+    async with engine.connect() as conn:
+        try:
+            logger.info("Attempting to create pgvector extension...")
+            await conn.execute(text("CREATE EXTENSION IF NOT EXISTS vector"))
+            await conn.commit()
+            logger.info("pgvector extension created successfully.")
+        except ProgrammingError as e:
+            logger.error(f"ProgrammingError: {e}")
+            raise RuntimeError(
+                "Failed to create pgvector extension. Make sure it's installed on the PostgreSQL server."
+            ) from e
+        except OperationalError as e:
+            logger.error(f"OperationalError: {e}")
+            raise RuntimeError(
+                "Failed to connect to the database. Check your connection settings."
+            ) from e
+        except Exception as e:
+            logger.error(f"Unexpected error: {e}")
+            raise
+    return engine
+
+
+@pytest_asyncio.fixture(scope="session")
+async def db_engine():
+    create_test_database(TEST_DB_URL)
+    engine = await setup_test_database(TEST_DB_URL)
+
+    async with engine.begin() as conn:
+        await conn.run_sync(Base.metadata.create_all)
+
+    yield engine
+
+    await engine.dispose()
+
+    drop_database(TEST_DB_URL)
+
+
+@pytest_asyncio.fixture(scope="function")
+async def db_session(db_engine):
+    """Create a database session for the scope of a single test function"""
+    Session = async_sessionmaker(bind=db_engine, expire_on_commit=False)
+    async with Session() as session:
+        yield session
+        await session.rollback()
+
+
+@pytest.fixture(scope="function")
+def client(db_session):
+    """Create a FastAPI TestClient for the scope of a single test function"""
+
+    async def override_get_db():
+        yield db_session
+
+    app.dependency_overrides[get_db] = override_get_db
+    with TestClient(app) as c:
+        yield c
+
+
+@pytest_asyncio.fixture(scope="function")
+async def test_data(db_session):
```

Comment:
shouldn't name non testing function with test_ prefix i dont think it will cause errors but might want to rename to example or something